crates/turborepo-cache/src/cache_archive/restore_regular.rs

```diff
@@ -0,0 +1,57 @@
+use std::{fs::OpenOptions, io, io::Read, path::Path};
+
+use tar::Entry;
+use turbopath::{AbsoluteSystemPath, AnchoredSystemPath, AnchoredSystemPathBuf};
+
+use crate::{
+    cache_archive::{restore::canonicalize_name, restore_directory::CachedDirTree},
+    CacheError,
+};
+
+pub fn restore_regular(
+    dir_cache: &mut CachedDirTree,
+    anchor: &AbsoluteSystemPath,
+    entry: &mut Entry<impl Read>,
+) -> Result<AnchoredSystemPathBuf, CacheError> {
+    let header = entry.header();
+    // Assuming this was a `turbo`-created input, we currently have an
+    // AnchoredUnixPath. Assuming this is malicious input we don't really care
+    // if we do the wrong thing.
+    let processed_name = canonicalize_name(&header.path()?)?;
+
+    // We need to traverse `processedName` from base to root split at
+    // `os.Separator` to make sure we don't end up following a symlink
+    // outside of the restore path.
+    dir_cache.safe_mkdir_file(anchor, &processed_name)?;
+
+    let resolved_path = anchor.resolve(&processed_name);
+    let mut open_options = OpenOptions::new();
+    open_options.write(true).truncate(true).create(true);
+
+    #[cfg(unix)]
+    {
+        use std::os::unix::fs::OpenOptionsExt;
+        open_options.mode(header.mode()?);
+    }
+
+    let mut file = open_options.open(resolved_path.as_path())?;
+    io::copy(entry, &mut file)?;
+
+    Ok(processed_name)
+}
+
+impl CachedDirTree {
+    pub fn safe_mkdir_file(
+        &mut self,
+        anchor: &AbsoluteSystemPath,
+        processed_name: &AnchoredSystemPath,
+    ) -> Result<(), CacheError> {
+        let is_root_file = processed_name.as_path().parent() == Some(Path::new("."));
```

Comment:
Drop `as_path()`.

Have you verified that `parent()` returns `"."`?

The docs say "This means it returns `Some("")` for relative paths with one component"